File: solutions/2024/11-Plutonian_Pebbles/s1.py
import os
current_dir = os.path.dirname(os.path.realpath(__file__))
# input_path = os.path.join(os.path.dirname(__file__), 'input_example')
input_path = os.path.join(os.path.dirname(__file__), 'input')

with open(input_path, 'r', encoding='utf8') as file:
    data = file.read().split('\n')
    data = list(map(int, data[0].split(' ')))
    
def solution_1(data):
    num_gens = 25
    gen = 0
    
    while gen < num_gens:
        gen += 1
        print(gen)
        
        new_data = []
        
        for stone in data:
            if stone == 0:
                new_data.append(1)
            elif len(str(stone)) % 2 == 0:
                _stone = str(stone)
                left = _stone[:len(_stone)//2]
                right = _stone[len(_stone)//2:]
                new_data.append(int(left))
                new_data.append(int(right))
            else:
                new_data.append(stone*2024)
                
            data = new_data
        
    # print(new_data)
    print(len(new_data))

# solution_1(data)

def solution_2(data):
    
    _data = {}
    for _ in data:
        if _ in _data:
            _data[_] +=1
        else:
            _data[_] = 1
    data = _data
    
    num_gens = 75
    gen = 0
    
    while gen < num_gens:
        # print(data)
        gen += 1
        
        new_data = {}
        
        for stone_value in data:
            if stone_value == 0:
                if 1 not in new_data:
                    new_data[1] = data[0]
                else:
                    new_data[1] += data[0]
                    
            elif len(str(stone_value)) % 2 == 0:
                _stone = str(stone_value)
                left = _stone[:len(_stone)//2]
                right = _stone[len(_stone)//2:]
                
                if int(left) not in new_data:
                    new_data[int(left)] = data[stone_value]
                else:
                    new_data[int(left)] += data[stone_value]
                
                if int(right) not in new_data:
                    new_data[int(right)] = data[stone_value]
                else:
                    new_data[int(right)] += data[stone_value]
                    
            else:
                if stone_value*2024 not in new_data:
                    new_data[stone_value*2024] = data[stone_value]
                else:
                    new_data[stone_value*2024] += data[stone_value]
                
        data = new_data
        
    solution = 0
    for _ in data:
        solution += data[_]
    print(solution)

solution_2(data)

File: solutions/2023/10-Pipe_Maze/s1.js
const utils = require('../../../templates/javascript/utils.js')
const { submit } = require('../../../lib/aoc-utils/utils.js')
const path = require('path')
const readFileSync = require('fs').readFileSync
const inputPath = path.join(__dirname, 'input')
const data = readFileSync(inputPath, 'utf8')
const lines = data.split(/\n/)

const dirs = {
	up: { x: 0, y: -1 },
	right: { x: 1, y: 0 },
	down: { x: 0, y: 1 },
	left: { x: -1, y: 0 }
};

const dirsOpp = {
	'{"x":0,"y":-1}': dirs.down,
	'{"x":1,"y":0}': dirs.left,
	'{"x":0,"y":1}': dirs.up,
	'{"x":-1,"y":0}': dirs.right
};

function getOppositeDirection(direction) {
	const dirString = JSON.stringify(direction);
	return dirsOpp[dirString];
}

const pipes = {
	"|": [dirs.up, dirs.down],
	"-": [dirs.left, dirs.right],
	"L": [dirs.up, dirs.right],
	"J": [dirs.up, dirs.left],
	"7": [dirs.left, dirs.down],
	"F": [dirs.right, dirs.down]
}

function isValidConnection(prevPipeDir, newPipeSymbol) {
	if (newPipeSymbol === '.') {
		return false
	}
	return pipes[newPipeSymbol].includes(getOppositeDirection(prevPipeDir))
}
function nextPipeDir(prevPipeDir, newPipeSymbol) {
	return pipes[newPipeSymbol].filter(dir => dir !== getOppositeDirection(prevPipeDir))[0]
}

function getNextIndex(x, y, dir) {
	const newX = x + dir.x
	const newY = y + dir.y

	if (newX >= minX && newX <= maxX && newY >= minY && newY <= maxY) {
		return { x: newX, y: newY }
	}
	return false
}

function findStartIndex(lines) {
	for (let y = 0; y < lines.length; y++) {
		for (let x = 0; x < lines[0].length; x++) {
			if (lines[y][x] === 'S') {
				return { x, y }
			}
		}
	}
}

const minX = 0
const minY = 0
const maxX = lines[0].length - 1
const maxY = lines.length - 1

function solution1(lines) {
	const startIndex = findStartIndex(lines)
	let currIndex = startIndex

	for (const assumedPipe of Object.keys(pipes)) {
		let prevPipeDir = pipes[assumedPipe][0]
		let steps = 0

		do {

			const nextIndex = getNextIndex(currIndex.x, currIndex.y, prevPipeDir)

			if (!nextIndex) {
				break
			}

			nextPipeSymbol = lines[nextIndex.y][nextIndex.x]
			if (nextPipeSymbol === 'S') {
				return Math.ceil(steps / 2)
			}

			if (!isValidConnection(prevPipeDir, nextPipeSymbol)) {
				break
			}


			newPipeDir = nextPipeDir(prevPipeDir, nextPipeSymbol)

			currIndex = nextIndex
			prevPipeDir = newPipeDir
			steps++


		} while (currIndex != startIndex)

	}

}




const SOLUTION_1 = solution1(lines)

module.exports = {}

// submit(SOLUTION_1)

console.log(`SOLUTION_1: ${SOLUTION_1}`)
// Example solution: 8